core/module.go

```diff
@@ -4,66 +4,568 @@ import (
+		if typeDef.AsObject.Valid {
+			typeDef.AsObject.Value.SourceModuleName = mod.Name()
+		}
+		typeDefs = append(typeDefs, typeDef)
+	}
+	for _, def := range mod.InterfaceDefs {
+		typeDef := def.Clone()
+		if typeDef.AsInterface.Valid {
+			typeDef.AsInterface.Value.SourceModuleName = mod.Name()
+		}
+		typeDefs = append(typeDefs, typeDef)
+	}
+	return typeDefs, nil
+}
+
+func (mod *Module) DependencySchemaIntrospectionJSON(ctx context.Context) (string, error) {
+	return mod.Deps.SchemaIntrospectionJSON(ctx)
+}
+
+func (mod *Module) ModTypeFor(ctx context.Context, typeDef *TypeDef, checkDirectDeps bool) (ModType, bool, error) {
+	var modType ModType
+	switch typeDef.Kind {
+	case TypeDefKindString, TypeDefKindInteger, TypeDefKindBoolean, TypeDefKindVoid:
+		modType = &PrimitiveType{typeDef}
+
+	case TypeDefKindList:
+		underlyingType, ok, err := mod.ModTypeFor(ctx, typeDef.AsList.Value.ElementTypeDef, checkDirectDeps)
+		if err != nil {
+			return nil, false, fmt.Errorf("failed to get underlying type: %w", err)
+		}
+		if !ok {
+			return nil, false, nil
+		}
+		modType = &ListType{
+			Elem:       typeDef.AsList.Value.ElementTypeDef,
+			Underlying: underlyingType,
+		}
+
+	case TypeDefKindObject:
+		if checkDirectDeps {
+			// check to see if this is from a *direct* dependency
+			depType, ok, err := mod.Deps.ModTypeFor(ctx, typeDef)
+			if err != nil {
+				return nil, false, fmt.Errorf("failed to get type from dependency: %w", err)
+			}
+			if ok {
+				return depType, true, nil
+			}
+		}
+
+		var found bool
+		// otherwise it must be from this module
+		for _, obj := range mod.ObjectDefs {
+			if obj.AsObject.Value.Name == typeDef.AsObject.Value.Name {
+				modType = &ModuleObjectType{
+					typeDef: obj.AsObject.Value,
+					mod:     mod,
+				}
+				found = true
+				break
+			}
+		}
+		if !found {
+			slog.Debug("module did not find object", "mod", mod.Name(), "object", typeDef.AsObject.Value.Name)
+			return nil, false, nil
+		}
+
+	case TypeDefKindInterface:
+		if checkDirectDeps {
+			// check to see if this is from a *direct* dependency
+			depType, ok, err := mod.Deps.ModTypeFor(ctx, typeDef)
+			if err != nil {
+				return nil, false, fmt.Errorf("failed to get interface type from dependency: %w", err)
+			}
+			if ok {
+				return depType, true, nil
+			}
+		}
+
+		var found bool
+		// otherwise it must be from this module
+		for _, iface := range mod.InterfaceDefs {
+			if iface.AsInterface.Value.Name == typeDef.AsInterface.Value.Name {
+				modType = &InterfaceType{
+					mod:     mod,
+					typeDef: iface.AsInterface.Value,
+				}
+				found = true
+				break
+			}
+		}
+		if !found {
+			slog.Debug("module did not find interface", "mod", mod.Name(), "object", typeDef.AsInterface.Value.Name)
+			return nil, false, nil
+		}
+
+	default:
+		return nil, false, fmt.Errorf("unexpected type def kind %s", typeDef.Kind)
+	}
+
+	if typeDef.Optional {
+		modType = &NullableType{
+			Elem:       typeDef.WithOptional(false),
```

Comment:
Is it supposed to be `typeDef.WithOptional(true)`? Or is this just preventing the case of like `Optional[Optional[T]]` and ensuring it's only one degree of `Optional`ity?

Reply:
Yeah the latter - had to double-check this one, but it's because our `TypeDefs` implement optional by just flipping a `bool` field, but in DagQL's sort-of-type-system it's handled through wrapping, with each concrete type always reporting its `*ast.Type` as `NonNull: true` and the nullable wrappers just flipping that to `false`.

I should rename `NullableType.Elem` to `Inner` or something - that's probably copypasta from `ListType` which has similar semantics, but "Elem" is short for "list element" in that case.